node/manager.go

```diff
@@ -0,0 +1,206 @@
+// This source code is provided 'as is' and no warranties are given as to title or non-infringement, merchantability
+// or fitness for purpose and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+// This source code is governed by Apache License 2.0 that can be found in the LICENSE file.
+
+package node
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/iotexproject/go-pkgs/crypto"
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-core/pkg/log"
+	"github.com/iotexproject/iotex-core/pkg/routine"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+	"github.com/iotexproject/iotex-core/pkg/version"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	transmitter interface {
+		BroadcastOutbound(context.Context, proto.Message) error
+		UnicastOutbound(context.Context, peer.AddrInfo, proto.Message) error
+	}
+
+	heightable interface {
+		TipHeight() uint64
+	}
+
+	privateKey interface {
+		PublicKey() crypto.PublicKey
+		Sign([]byte) ([]byte, error)
+	}
+
+	// Info node infomation
+	Info struct {
+		Version   string
+		Height    uint64
+		Timestamp time.Time
+		Address   string
+	}
+
+	// DelegateManager manage delegate node info
+	DelegateManager struct {
+		cfg          Config
+		nodeMap      map[string]Info
+		nodeMapMutex sync.Mutex
+		broadcaster  *routine.RecurringTask
+		transmitter  transmitter
+		heightable   heightable
+		privKey      privateKey
+	}
+)
+
+// NewDelegateManager new delegate manager
+func NewDelegateManager(cfg *Config, t transmitter, h heightable, privKey privateKey) *DelegateManager {
+	dm := &DelegateManager{
+		cfg:         *cfg,
+		nodeMap:     make(map[string]Info),
+		transmitter: t,
+		heightable:  h,
+		privKey:     privKey,
+	}
+
+	dm.broadcaster = routine.NewRecurringTask(func() {
+		// delegates and nodes who are turned on will broadcast
+		if !dm.isDelegate() && dm.cfg.OnlyDelegateBroadcast {
+			log.L().Debug("delegate manager general node disabled node info broadcast")
+			return
+		}
+
+		if err := dm.BroadcastNodeInfo(context.Background()); err != nil {
+			log.L().Error("delegate manager broadcast node info failed", zap.Error(err))
+		}
+	}, cfg.BroadcastNodeInfoInterval)
+	return dm
+}
+
+// Start start delegate broadcast task
+func (dm *DelegateManager) Start(ctx context.Context) error {
+	if dm.broadcaster != nil {
+		return dm.broadcaster.Start(ctx)
+	}
+	return nil
+}
+
+// Stop stop delegate broadcast task
+func (dm *DelegateManager) Stop(ctx context.Context) error {
+	if dm.broadcaster != nil {
+		return dm.broadcaster.Stop(ctx)
+	}
+	return nil
+}
+
+// HandleNodeInfo handle node info message
+func (dm *DelegateManager) HandleNodeInfo(ctx context.Context, peerID string, msg *iotextypes.ResponseNodeInfoMessage) {
```

Comment:
`peerID` is not used in this func?
should we also store `peerID` together with NodeInfo, so user can query a node's info either by address `io1xxxxx` or by the peerID `12D3Kooxxx`?

Reply:
good suggestion, the `peerID` is necessary if we want to send unicast request msg (e.g., i want to sync a node info when i find it's node info is out of date in local memory)